samples/boards/nrf52/power_mgr/src/main.c

```diff
@@ -0,0 +1,330 @@
+/*
+ * Copyright (c) 2016 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+#include <power.h>
+#include <soc_power.h>
+#include <misc/printk.h>
+#include <string.h>
+#include <board.h>
+#include <device.h>
+#include <gpio.h>
+
+#define SECONDS_TO_SLEEP	1
+
+/* In Tickless Kernel mode, time is passed in milliseconds instead of ticks */
+#ifdef CONFIG_TICKLESS_KERNEL
+#define TICKS_TO_SECONDS_MULTIPLIER 1000
+#define TIME_UNIT_STRING "milliseconds"
+#else
+#define TICKS_TO_SECONDS_MULTIPLIER CONFIG_SYS_CLOCK_TICKS_PER_SEC
+#define TIME_UNIT_STRING "ticks"
+#endif
+#if defined(CONFIG_SYS_POWER_DEEP_SLEEP)
+#define MAX_POWER_SAVING_STATE 3
+#else
+#define MAX_POWER_SAVING_STATE 2
+#endif
+#define NO_POWER_SAVING_STATE 0
+
+#define MIN_TIME_TO_SUSPEND	((SECONDS_TO_SLEEP * \
+				  TICKS_TO_SECONDS_MULTIPLIER) - \
+				  (TICKS_TO_SECONDS_MULTIPLIER / 2))
+#define GPIO_CFG_SENSE_LOW (GPIO_PIN_CNF_SENSE_Low << GPIO_PIN_CNF_SENSE_Pos)
+
+/* To Create Ordered List of devices */
+static void create_device_list(void);
+/* Suspend Devices based on Ordered List */
+static void suspend_devices(void);
+/* Resume Devices */
+static void resume_devices(void);
+static struct device *device_list;
+static int device_count;
+static int post_ops_done = 1;
+
+/*
+ * Example ordered list to store devices on which
+ * device power policies would be executed.
+ */
+#define DEVICE_POLICY_MAX 15
+static char device_ordered_list[DEVICE_POLICY_MAX];
+static char device_retval[DEVICE_POLICY_MAX];
+static int pm_state;
+static struct device *gpiob;
+
+/* change this to use another GPIO port */
+#define PORT    SW0_GPIO_NAME
+#define PIN	SW0_GPIO_PIN
+#define LOW	0
+
+#define DEMO_DESCRIPTION	\
+	"Demo Description\n"	\
+	"Application creates Idleness, Due to which System Idle Thread is\n"\
+	"scheduled and it enters into various Low Power States.\n"\
+	"Press Button 1 on Board to wake device from Low Power State.\n"
+
+extern int nrf_gpiote_interrupt_enable(uint32_t mask);
+
+void configure_ram_retention(void)
+{
+	/* Place Holder for RAM retention */
+	/* Turn Off the RAM Blocks which are not needed */
+}
+
+/* Initialize and configure GPIO */
+static void gpio_setup(void)
+{
+
+	gpiob = device_get_binding(PORT);
+	gpio_pin_configure(gpiob, PIN, GPIO_DIR_IN
+				| GPIO_PUD_PULL_UP
+				| GPIO_CFG_SENSE_LOW);
+	/* Enable GPIOTE Port Event */
+	nrf_gpiote_interrupt_enable(GPIOTE_INTENSET_PORT_Msk);
```

Comment:
I prefer not to hide the EVENT clear inside this function, refer to my other comments related to use if this function.